File: packages/televault-core/src/vaultEntry.js
/**
 * @typedef {Object} VaultEntry
 * @property {number} messageId
 * @property {string} path - File: '/a/b.pdf'. Dir marker: '/a/b/'.
 * @property {number} size
 * @property {string} sha256
 * @property {Date} mtime
 * @property {string[]} tags
 * @property {string|null} [localPath]
 */

/** @returns {VaultEntry} */
function createVaultEntry({ messageId, path, size, sha256, mtime, tags = [], localPath = null }) {
  return { messageId, path, size, sha256, mtime, tags, localPath };
}

/** @returns {VaultEntry} */
function dirMarker({ messageId, path, tags = [] }) {
  if (!path.endsWith('/')) {
    throw new Error('dir marker path must end with /');
  }
  return createVaultEntry({
    messageId,
    path,
    size: 0,
    sha256: '',
    mtime: new Date(),
    tags,
  });
}

/** @param {VaultEntry} entry */
function isDir(entry) {
  return entry.path.endsWith('/');
}

/** @param {VaultEntry} entry */
function entryName(entry) {
  const p = isDir(entry) ? entry.path.slice(0, -1) : entry.path;
  return p.slice(p.lastIndexOf('/') + 1);
}

/** @param {VaultEntry} entry */
function entryParent(entry) {
  const p = isDir(entry) ? entry.path.slice(0, -1) : entry.path;
  const i = p.lastIndexOf('/');
  return i <= 0 ? '/' : `${p.slice(0, i + 1)}`;
}

/** @param {VaultEntry} entry */
function copyEntry(entry, { path, tags, localPath, messageId } = {}) {
  return createVaultEntry({
    messageId: messageId ?? entry.messageId,
    path: path ?? entry.path,
    size: entry.size,
    sha256: entry.sha256,
    mtime: entry.mtime,
    tags: tags ?? entry.tags,
    localPath: localPath ?? entry.localPath,
  });
}

module.exports = {
  createVaultEntry,
  dirMarker,
  isDir,
  entryName,
  entryParent,
  copyEntry,
};

File: televault/lib/models/vault_entry.dart
class VaultEntry {
  final int messageId;

  /// File: '/a/b.pdf'. Thư mục marker: '/a/b/'.
  final String path;
  final int size;
  final String sha256;
  final DateTime mtime;
  final List<String> tags;

  /// Điền sau từ IndexDB — đường dẫn file đã cache local (null nếu chưa tải).
  final String? localPath;

  const VaultEntry({
    required this.messageId,
    required this.path,
    required this.size,
    required this.sha256,
    required this.mtime,
    this.tags = const [],
    this.localPath,
  });

  factory VaultEntry.dirMarker({required int messageId, required String path, List<String> tags = const []}) {
    assert(path.endsWith('/'));
    return VaultEntry(
      messageId: messageId,
      path: path,
      size: 0,
      sha256: '',
      mtime: DateTime.now().toUtc(),
      tags: tags,
    );
  }

  bool get isDir => path.endsWith('/');

  String get name {
    final p = isDir ? path.substring(0, path.length - 1) : path;
    return p.substring(p.lastIndexOf('/') + 1);
  }

  /// Thư mục cha, luôn kết thúc '/'. Ví dụ '/a/b.pdf' -> '/a/', '/a/' -> '/'.
  String get parent {
    final p = isDir ? path.substring(0, path.length - 1) : path;
    final i = p.lastIndexOf('/');
    return i <= 0 ? '/' : p.substring(0, i + 1);
  }

  VaultEntry copyWith({String? path, List<String>? tags, String? localPath, int? messageId}) {
    return VaultEntry(
      messageId: messageId ?? this.messageId,
      path: path ?? this.path,
      size: size,
      sha256: sha256,
      mtime: mtime,
      tags: tags ?? this.tags,
      localPath: localPath ?? this.localPath,
    );
  }
}
